packages/aws-cdk/lib/api/toolkit-info.ts

```diff
@@ -60,17 +65,31 @@ export class ToolkitInfo {
       return { filename, key, changed: false };
     }
 
-    debug(`${url}: uploading`);
-    await s3.putObject({
-      Bucket: bucket,
-      Key: key,
-      Body: data,
-      ContentType: props.contentType
-    }).promise();
-
-    debug(`${url}: upload complete`);
+    const uploaded = { filename, key, changed: true };
+
+    // Upload if it's new or server-side copy if it was already uploaded previously
+    const previous = this.previousUploads[hash];
+    if (previous) {
```

Comment:
Why not just reuse the same file that was already uploaded?

Reply:
We're organizing the data in folders, by `Asset` instance.

This is currently so that we can give permissions on the folder to constructs that require access to the asset, so they will be able to download all versions of it, so that even if we upload a new version and the Execution Role is updated to only allow downloading the latest version of the asset, old runs/definitions will not stop working. This is not required for Lambdas which have a point-in-time handover but it *is* required for CodeBuild script assets, for example, which download the file at an arbitrary later point in time.

If we reused the same file between multiple assets, we'd either have to fall back to "only permissions to the latest version of the asset" or "permission to all assets in the bucket".